### ISSUE TITLE
Some pathological cases

### ISSUE THREAD

**Issue by joliss** [pornel/pngquant#201]
I found two related pathological cases that I'd like to share with you. In both cases I'll show how manual optimization yields better perceptual quality than pngquant's algorithm. If you want to experiment for yourself, run `build.sh` in the [pngquant-issue](https://github.com/joliss/pngquant-issue) repo.

## (1) Card back

If we quantize the following image with 38 colors, we get a file that's mostly very close to the original, but exhibits color drift in the blue card-back pattern at the bottom:

**Original:** [1-original.png](https://raw.githubusercontent.com/joliss/pngquant-issue/master/out/1-original.png)

**pngquant's 38-color version:** [1-or8.png](https://raw.githubusercontent.com/joliss/pngquant-issue/master/out/1-or8.png)

We can force better allocation for the palette entries by quantizing the [base image](https://raw.githubusercontent.com/joliss/pngquant-issue/master/base.png) with 30 colors, the [card-back image](https://raw.githubusercontent.com/joliss/pngquant-issue/master/cardback.png) with 9 colors, and compositing the two into one 38 colors image:

**Hand-optimized 38-color version:** [1-composite-or8.png](https://raw.githubusercontent.com/joliss/pngquant-issue/master/out/1-composite-or8.png)

The root cause here seems to be that mean squared error doesn't account for bias in some part of the image - if the error was randomly distributed, it would average out, but here the error values on the card-back texture all point in the same direction, causing the color to shift.

## (2) Reload button

If we quantize the following image with 48 colors, we get a file that's again very close to the original, except for the reload button at the bottom, where it seems surprisingly far off from the original.

**Original:** [2-original.png](https://raw.githubusercontent.com/joliss/pngquant-issue/master/out/2-original.png)

**pngquant's 48-color version (66 KB):** [2-or8.png](https://raw.githubusercontent.com/joliss/pngquant-issue/master/out/2-or8.png)

Like in the previous case, the underlying issue seems to be that an entire area of pixels is systematically drifting away from the original. (I produced a dithered version, [2-fs8.png](https://raw.githubusercontent.com/joliss/pngquant-issue/master/out/2-fs8.png), but I think it's a toss-up.) It seems clear that perceptually, the gradient colors in the reload button need more palette entries allocated to them. We can produce a hand-optimized version by quantizing the [base image](https://raw.githubusercontent.com/joliss/pngquant-issue/master/base.png) with 40 colors, the [reload image](https://raw.githubusercontent.com/joliss/pngquant-issue/master/reload.png) with 9 colors, and compositing the two into one 48 colors image:

**Hand-optimized 48-color version (64 KB):** [2-composite-a-or8.png](https://raw.githubusercontent.com/joliss/pngquant-issue/master/out/2-composite-a-or8.png)

What we *really* care about, of course, is not the number of colors per se, but the file size of the resulting PNG. If instead of saying, "give me a file with a *n* colors", we restate the problem as "give me a file *n* KB in size", we have more freedom to spend any number of colors (up to 256 anyway) on a limited part of the image that benefits from it, such as the reload button in this example. If we quantize the [base image](https://raw.githubusercontent.com/joliss/pngquant-issue/master/base.png) with 46 colors and the [reload image](https://raw.githubusercontent.com/joliss/pngquant-issue/master/reload.png) with 85 colors, we get a much superior image with the exact same file size as pngquant's version:

**Hand-optimized 130-color version (66 KB):** [2-composite-b-or8.png](https://raw.githubusercontent.com/joliss/pngquant-issue/master/out/2-composite-b-or8.png)

-------------

So to summarize, it seems like we're running into two issues here:

1. *Mean squared error* doesn't always measure psychovisual quality well.
2. We may get better results if we treat file size, rather than color count, as the optimization constraint within which we maximize quality.

Both of these are pretty hard problems, but if you ever want to dig into them, I'm hoping that the images I've provided might come in handy!

**Comment by joliss** [comment]
Addendum: Usually size is a monotonically increasing function of color count. So why should we care about optimizing for size, when we could simply treat color count as a proxy? I came up with two reasons:

* If we optimize a file for size, we might use a high number of colors for a small part of the image (e.g. a gradient), but choose not to use of those colors in the rest of the image. If our optimization goal is the number of colors, this makes no sense: Why wouldn't we use those colors everywhere in the image? But if our goal is file size, it does makes sense: The rest of the image might not benefit as much from more colors, and it will compress better if we restrict ourselves to a smaller palette. In other words, by treating file size as a goal we can come up with a better algorithm.

* If we can ask for a file size, like "give me a 70 KB file" or "give me an 80 KB file", a good algorithm might produce a file with 256 colors in both cases. If we can only ask for a color count, then once we're maxed out at 256 colors we can only get whatever size the algorithm chooses for us (70 KB *or* 80 KB).

**Comment by pornel** [comment]
In pngquant I'm optimizing for MSE of my made-up "perceptual weight" of colors, where pixels in "noisy" areas get less weight than in smooth areas.

I'm not optimizing for anything more realistic because better metrics are way more expensive to compute, to the point of being infeasible.

File size is not a quality metric, so I still need some quality metric to combine it with (to optimize for "best quality under xKB" or "best quality/filesize ratio"). 

Gzipped file size is also a pain to work with, because it's slow and not monotonic. More precise colors can lead to less dithering, and smaller file size, because dithering compresses poorly.

I can measure MSE of a palette against histogram cheaply, but e.g. an SSIM metric would require remapping and comparison against full reference, which is orders of magnitude slower. 

MSE is also simple and fast enough to be used in gradient descent. Feedback from SSIM is fuzzy, and for really good results requires complicated pooling. But then the score can't be used in a straightforward way to adjust histogram weights in the gradient descent.

**Comment by joliss** [comment]
Very interesting. I suppose there might not be a satisfactory solution then.

I'll go ahead and close this. (Unless you want to keep it open of course.)

It's not clear to me that there's a way to improve the situation with bias or banding without sacrificing performance, but if you end up working on the perceptual weight function again and can think of something, perhaps the test images I've posted may come in useful somehow.

**Comment by pornel** [comment]
Thanks for the test cases.

I know they're handled well by the [LBG](https://en.wikipedia.org/wiki/Linde%E2%80%93Buzo%E2%80%93Gray_algorithm) algorithm:

![lbg](http://www.data-compression.com/lbgvq.gif)

but it again is order of magnitude more expensive. It may be worth to investigate some kind of approximation/hybrid version of it.
